File: Aditya Vardhan Singh/Other Problems/Day 12/2095 Delete the Middle Node of a Linked List/C++/2095 Delete the Middle Node of a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        /* edge cases */
        if (!head->next) {
            // single node only
            delete(head);
            return nullptr;
        }
        if (!head->next->next) {
            // two nodes only
            delete(head->next);
            head->next = nullptr;
            return head;
        }
        /* ends */

        /* find middle of list */
        ListNode * slow, * fast;
        slow = fast = head;

        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        /* ends */

        /* remove middle node */
        slow->val = slow->next->val;
        ListNode *temp = slow->next; // node to be deleted
        slow->next = slow->next->next;
        temp->next = nullptr;
        delete(temp);
        /* ends */

        return head;
    }
};

File: Aditya Vardhan Singh/Other Problems/Day 12/2095 Delete the Middle Node of a Linked List/Python/2095 Delete the Middle Node of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head.next is None: # one node only
            return None
        if head.next.next is None: # two nodes only
            head.next = None
            return head

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        slow.val = slow.next.val
        slow.next = slow.next.next
        
        return head